Added a junit test point for windowed functions and partitioned tables.

BillWhite opened the pull request:
We test both partitioned tables in windowed partition by expressions and
in windowed order by expressions.

wolffcm commented:
In all these examples, the PARTITION BY clause contains just the partition key for the table.  What happens when you PARTITION BY something that is not?

In this case we need to do some work on the coordinator.  Do we get that right?

wolffcm commented:
Also, do we have planner tests that lock down the plans generated when we do RANK on a partitioned table where the PARTITION BY clause is not a table's partition key?

wolffcm commented:
That's all my comments.  I think we need some more test cases here.

BillWhite commented:
I added some more regression tests with mixed partition by and partitioned tables.  I also added some planner tests, but the plans for these will change quite significantly in the future.  Most of the work happens on the coordinator fragment anyway.

wolffcm commented:
Thanks Bill, I'll have a look when I'm done reviewing Ethan's changes.

wolffcm commented on tests/frontend/org/voltdb/planner/TestWindowedFunctions.java:
```diff
@@ -238,6 +239,59 @@ private void validateQueryWithSubquery(String windowedQuery) {
+            assertTrue(child instanceof OrderByPlanNode);
+        }
+        child = child.getChild(0);
+        assertTrue(child instanceof PartitionByPlanNode);
```
So we always put the PartitionByPlanNode on the coordinator?  When we're doing PARTITION BY on a table's partition key, then we could do the work on the distributed fragment, and then concatenate the result from all the partitions.

wolffcm replied:
We do something similar to this when we handle aggregates: if the GROUP BY the partition key, then we can just concatenate the aggregations produced by the distributed fragment on the coordinator.  I'm kind of surprised we're not doing it here.

BillWhite replied:
I had thought optimizations like this were part of the follow on ticket.

wolffcm replied:
I see.  Using indexes to avoid sorting and pushing down PartitionByPlanNode into distributed fragments are really two separate optimizations that can be applied independently.  I think it probably needs a separate ticket.

Not executing PartitionByPlanNode in the distributed fragments is not as big a deal as it is for aggregates, but still worth doing.

wolffcm commented on tests/frontend/org/voltdb/planner/testplans-windowingfunctions-ddl.sql:
```diff
@@ -0,0 +1,38 @@
+);
+PARTITION TABLE AAA_PA ON COLUMN A;
+
+CREATE TABLE AAA_PB (
```
This table seems unused?

wolffcm commented on tests/frontend/org/voltdb/regressionsuites/TestWindowedAggregateSuite.java:
```diff
@@ -93,6 +93,14 @@ static private void setupSchema(VoltProjectBuilder project) throws IOException {
+
                 + "CREATE TABLE T_PB (\n"
                 + "  A INTEGER NOT NULL,"
                 + "  B INTEGER NOT NULL,"
```
What does the table name `T_PB` mean?  Is it supposed to mean that the table is partitioned by column `B`?

wolffcm commented on tests/frontend/org/voltdb/regressionsuites/TestWindowedAggregateSuite.java:
```diff
@@ -259,7 +272,7 @@ public void testRankWithString() throws Exception {
             assertEquals(expected[rowIdx][colB], vt.getLong(1));
             assertEquals(Long.toString(expected[rowIdx][colC], 10), vt.getString(2));
-            assertEquals(expected[rowIdx][colR1], vt.getLong(3));
+            assertEquals(expected[rowIdx][colR_A], vt.getLong(3));
```
Do we have any tests where you do a PARTITION BY a string column?

BillWhite replied:
I added one of these.

wolffcm commented on tests/frontend/org/voltdb/regressionsuites/TestWindowedAggregateSuite.java:
```diff
@@ -337,7 +351,29 @@ public void testRankPartitionedTable() throws Exception {
+            assertEquals(msg, expected[rowIdx][colA],    vt.getLong(0));
+            assertEquals(msg, expected[rowIdx][colB],    vt.getLong(1));
+            assertEquals(msg, expected[rowIdx][colC],    vt.getLong(2));
+            assertEquals(msg, expected[rowIdx][colR_AA], vt.getLong(3));
```
Are there any tests where we are selecting from a partitioned table, but the table's partition key does not appear at all in the PARTITION BY clause?  Maybe I'm missing it, but I don't see it here, since T_PC is actually partitioned by column A.

BillWhite replied:
Sorry.  T_PC should have been partitioned on C.  I fixed this.

wolffcm commented:
That's all my comments.

wolffcm commented:
The latest commit looks good.
